File: packages/engine/src/liveEffects/dither.ts
/**
 * Live non-destructive dithering kernel.
 *
 * Algorithms: Floyd-Steinberg, Atkinson, Jarvis-Judice-Ninke, Stucki, Sierra,
 * Bayer (2/4/8 ordered), and blue-noise. Error diffusion is inherently
 * sequential; it runs on the CPU in a single row-major pass (serpentine
 * optional) — the same deterministic output the export path produces.
 *
 * Anchoring: when a `coordSpace` is supplied, the pattern phase is anchored
 * to document coordinates so panning/zooming never makes the pattern
 * "swim" relative to layer content. Without it, the pattern anchors to the
 * surface's own pixel grid.
 *
 * Alpha: dithered colour is never applied to pixels whose alpha falls below
 * `alphaCutoff` (they stay transparent), and the alpha channel itself is
 * never modified by dithering.
 */

import {
  bayerThresholdMatrix,
  KERNELS,
  type DitherAlgorithm as TechDitherAlgorithm,
} from '../exportPipeline/dither';
import type { ColorMetric, PaletteColor } from './paletteCore';
import { buildPaletteLookup, sanitizePalette } from './paletteCore';
import { hash2, seeded01 } from './prng';

export type DitherAlgorithm =
  | 'floyd-steinberg'
  | 'atkinson'
  | 'jarvis-judice-ninke'
  | 'stucki'
  | 'sierra'
  | 'bayer'
  | 'blue-noise';

export type DitherPaletteMode = 'none' | 'levels' | 'custom';

export interface DitherParams {
  algorithm: DitherAlgorithm;
  paletteMode: DitherPaletteMode;
  /** Bits per channel when paletteMode === 'levels' (1..8). */
  levels: number;
  /** Explicit palette when paletteMode === 'custom'. */
  colors: readonly (readonly number[])[];
  metric: ColorMetric;
  serpentine: boolean;
  /** 0..1 error/pattern strength. */
  strength: number;
  /** Bayer matrix size (2 | 4 | 8). */
  bayerSize: number;
  /** Pattern cell size in document pixels (>= 1). */
  cellSize: number;
  /** Pixels below this alpha are forced fully transparent. 0..1 */
  alphaCutoff: number;
  /** Deterministic seed. */
  seed: number;
}

export interface CoordSpace {
  /** Device pixels per document pixel (camera zoom x dpr). */
  scale: number;
  /** Device x of the document origin (camera transform e). */
  originX: number;
  /** Device y of the document origin (camera transform f). */
  originY: number;
  /** Device x of the rendered region origin (the backdrop capture offset). */
  regionX: number;
  /** Device y of the rendered region origin (the backdrop capture offset). */
  regionY: number;
}

/** Document coordinate of a local (region-space) pixel given a CoordSpace. */
export function docCoordOf(x: number, y: number, coordSpace: CoordSpace): { x: number; y: number } {
  const scale = coordSpace.scale > 0 ? coordSpace.scale : 1;
  return {
    x: (coordSpace.regionX + x - coordSpace.originX) / scale,
    y: (coordSpace.regionY + y - coordSpace.originY) / scale,
  };
}

const LUMINANCE: readonly [number, number, number] = [0.2126, 0.7152, 0.0722];

/** Apply the dither effect in place (mutates data, returns same ImageData). */
export function applyDither(
  imageData: ImageData,
  params: DitherParams,
  coordSpace?: CoordSpace,
): ImageData {
  const { data, width: w, height: h } = imageData;
  if (w === 0 || h === 0) return imageData;

  const algorithm = params.algorithm ?? 'floyd-steinberg';
  const strength = clamp01(params.strength ?? 1);
  const alphaCutoff = clamp01(params.alphaCutoff ?? 0);
  const serpentine = params.serpentine ?? true;
  const seed = Math.round(params.seed ?? 0) >>> 0;

  const scale = coordSpace && coordSpace.scale > 0 ? coordSpace.scale : 1;
  const cell = Math.max(1, params.cellSize ?? 1) * scale;
  const phaseX = coordSpace ? coordSpace.regionX : 0;
  const phaseY = coordSpace ? coordSpace.regionY : 0;
  const originX = coordSpace ? coordSpace.originX : 0;
  const originY = coordSpace ? coordSpace.originY : 0;

  if (algorithm === 'bayer' || algorithm === 'blue-noise') {
    applyOrdered(
      data,
      w,
      h,
      { ...params, algorithm },
      { strength, alphaCutoff, cell, phaseX, phaseY, originX, originY, scale, seed },
    );
    return imageData;
  }

  if (strength <= 0) return imageData;

  // 'none' palette mode = no quantization = no dithering at all. Error
  // diffusion requires a quantization target to diffuse error towards.
  if ((params.paletteMode ?? 'levels') === 'none') return imageData;

  // Palette-based error diffusion: quantize to nearest palette colour per
  // channel; otherwise quantize to levels.
  let palette: PaletteColor[] = [];
  let lookup: { find(r: number, g: number, b: number): PaletteColor } | null = null;
  if (params.paletteMode === 'custom') {
    palette = sanitizePalette(params.colors ?? []);
    if (palette.length > 0) {
      lookup = buildPaletteLookup(palette, params.metric ?? 'rgb');
    }
  }
  const levels = Math.max(1, Math.min(8, Math.round(params.levels ?? 4)));
  const step = 1 / ((1 << levels) - 1);

  const kernelName =
    algorithm === 'sierra' ? 'jarvis-judice-ninke' : (algorithm as TechDitherAlgorithm);
  const kernel = KERNELS[kernelName as keyof typeof KERNELS] ?? KERNELS['floyd-steinberg'];

  const errR = new Float32Array(w * h);
  const errG = new Float32Array(w * h);
  const errB = new Float32Array(w * h);

  for (let y = 0; y < h; y += 1) {
    const ltr = !serpentine || y % 2 === 0;
    for (let sx = 0; sx < w; sx += 1) {
      const x = ltr ? sx : w - 1 - sx;
      const o = (y * w + x) * 4;
      const a = data[o + 3]!;
      if (a / 255 < alphaCutoff || a === 0) {
        continue;
      }
      const eo = y * w + x;
      const r = clamp01(data[o]! / 255 + (errR[eo] ?? 0) * strength);
      const g = clamp01(data[o + 1]! / 255 + (errG[eo] ?? 0) * strength);
      const b = clamp01(data[o + 2]! / 255 + (errB[eo] ?? 0) * strength);

      let qr: number;
      let qg: number;
      let qb: number;
      if (lookup) {
        const c = lookup.find(Math.round(r * 255), Math.round(g * 255), Math.round(b * 255));
        qr = c[0] / 255;
        qg = c[1] / 255;
        qb = c[2] / 255;
      } else {
        qr = Math.round(r / step) * step;
        qg = Math.round(g / step) * step;
        qb = Math.round(b / step) * step;
      }
      data[o] = Math.round(clamp01(qr) * 255);
      data[o + 1] = Math.round(clamp01(qg) * 255);
      data[o + 2] = Math.round(clamp01(qb) * 255);

      const er = (r - qr) * strength;
      const eg = (g - qg) * strength;
      const eb = (b - qb) * strength;
      for (const entry of kernel) {
        const nx = ltr ? x + entry.dx : x - entry.dx;
        const ny = y + entry.dy;
        if (nx < 0 || nx >= w || ny >= h) continue;
        const ne = ny * w + nx;
        errR[ne] = (errR[ne] ?? 0) + er * entry.weight;
        errG[ne] = (errG[ne] ?? 0) + eg * entry.weight;
        errB[ne] = (errB[ne] ?? 0) + eb * entry.weight;
      }
    }
  }
  return imageData;
}

interface OrderedState {
  strength: number;
  alphaCutoff: number;
  cell: number;
  phaseX: number;
  phaseY: number;
  originX: number;
  originY: number;
  scale: number;
  seed: number;
}

function applyOrdered(
  data: Uint8ClampedArray,
  w: number,
  h: number,
  params: DitherParams,
  st: OrderedState,
): void {
  const levels = Math.max(1, Math.min(8, Math.round(params.levels ?? 4)));
  const step = 1 / ((1 << levels) - 1);
  let palette: PaletteColor[] = [];
  let lookup: { find(r: number, g: number, b: number): PaletteColor } | null = null;
  if (params.paletteMode === 'custom') {
    palette = sanitizePalette(params.colors ?? []);
    if (palette.length > 0) {
      lookup = buildPaletteLookup(palette, params.metric ?? 'rgb');
    }
  }
  const bayer =
    params.algorithm === 'bayer'
      ? bayerThresholdMatrix(Math.max(2, Math.min(8, params.bayerSize ?? 4)))
      : null;

  for (let y = 0; y < h; y += 1) {
    for (let x = 0; x < w; x += 1) {
      const o = (y * w + x) * 4;
      const a = data[o + 3]!;
      if (a / 255 < st.alphaCutoff || a === 0) continue;
      // Document-anchored phase: the cell index derives from the document
      // coordinate of the pixel, so the pattern is invariant under pan/zoom
      // (a fixed cell in doc px covers a fixed set of doc pixels).
      const invScale = 1 / (st.scale > 0 ? st.scale : 1);
      const docX = (st.phaseX + x - st.originX) * invScale;
      const docY = (st.phaseY + y - st.originY) * invScale;
      const cellSizeDoc = st.cell * invScale;
      const cellIdxX = Math.floor(docX / cellSizeDoc);
      const cellIdxY = Math.floor(docY / cellSizeDoc);
      let threshold: number;
      if (bayer) {
        const size = Math.sqrt(bayer.length);
        const t = (bayer[(cellIdxY % size) * size + (cellIdxX % size)]! + 0.5) / (size * size);
        threshold = (t - 0.5) * st.strength * step * 1.5;
      } else {
        threshold = (hash2(cellIdxX, cellIdxY, st.seed) - 0.5) * st.strength * step;
      }
      const r = clamp01(data[o]! / 255 + threshold);
      const g = clamp01(data[o + 1]! / 255 + threshold);
      const b = clamp01(data[o + 2]! / 255 + threshold);
      if (lookup) {
        const c = lookup.find(Math.round(r * 255), Math.round(g * 255), Math.round(b * 255));
        data[o] = c[0];
        data[o + 1] = c[1];
        data[o + 2] = c[2];
      } else {
        data[o] = quantizeByte(r, step);
        data[o + 1] = quantizeByte(g, step);
        data[o + 2] = quantizeByte(b, step);
      }
    }
  }
}

function quantizeByte(v: number, step: number): number {
  return Math.round(clamp01(Math.round(v / step) * step) * 255);
}

export function lumaOf(r: number, g: number, b: number): number {
  return (LUMINANCE[0] * r + LUMINANCE[1] * g + LUMINANCE[2] * b) / 255;
}

/** Deterministic seed sequence for UI "shuffle" buttons. */
export function nextDeterministicSeed(seed: number): number {
  return Math.round(seeded01(seed + 0x9e3779b9) * 4294967295);
}

function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}

File: crates/varve-effects/src/dither.rs
//! Dither kernel — faithful port of `packages/engine/src/liveEffects/dither.ts`.
//!
//! Algorithms: Floyd-Steinberg, Atkinson, Jarvis-Judice-Ninke, Stucki, Sierra,
//! Bayer (2/4/8 ordered), and blue-noise. Error diffusion is inherently
//! sequential; it runs in a single row-major pass (serpentine optional).
//!
//! Anchoring: when a `coordSpace` with `scale > 0` is supplied, the ordered
//! pattern phase is anchored to document coordinates so panning/zooming never
//! makes the pattern "swim" relative to layer content. Without it, the
//! pattern anchors to the surface's own pixel grid.
//!
//! Alpha: dithered colour is never applied to pixels whose alpha falls below
//! `alphaCutoff`, and the alpha channel itself is never modified.

use crate::palette_core::{self, ColorMetric, PaletteLookup};
use crate::prng::hash2;
use crate::{clamp01, clamp_byte, js_round, CoordSpace, Params};

/// Document coordinate of a local (region-space) pixel given a CoordSpace —
/// port of `docCoordOf` in dither.ts. Returns `(x, y)`.
pub fn doc_coord_of(x: f64, y: f64, coord_space: CoordSpace) -> (f64, f64) {
    let scale = if coord_space.scale > 0.0 {
        coord_space.scale
    } else {
        1.0
    };
    (
        (coord_space.region_x + x - coord_space.origin_x) / scale,
        (coord_space.region_y + y - coord_space.origin_y) / scale,
    )
}

/// Recursively construct a Bayer threshold matrix — port of
/// `bayerThresholdMatrix` in exportPipeline/dither.ts.
///
/// The TS uses `const n = matrix.length` (the whole array, not the row
/// count), so on the second iteration the source reads are out of bounds:
/// `Uint16Array` yields `undefined`, arithmetic produces `NaN`, and the
/// `Uint16Array` assignment zeroes it. This port reproduces that NaN-to-zero
/// behaviour; a size-4 request therefore yields a 64-entry matrix (8x8).
fn bayer_threshold_matrix(size: f64) -> Vec<u16> {
    let mut matrix = vec![0u16];
    while (matrix.len() as f64) < size * size {
        let n = matrix.len();
        let next_n = n * 2;
        let mut next = vec![0u16; n * n * 4];
        for y in 0..n {
            for x in 0..n {
                let v: f64 = matrix.get(y * n + x).map(|v| *v as f64).unwrap_or(f64::NAN);
                next[y * next_n + x] = (v * 4.0) as u16;
                next[y * next_n + x + n] = (v * 4.0 + 2.0) as u16;
                next[(y + n) * next_n + x] = (v * 4.0 + 3.0) as u16;
                next[(y + n) * next_n + x + n] = (v * 4.0 + 1.0) as u16;
            }
        }
        matrix = next;
    }
    matrix
}

struct KernelEntry {
    dx: i64,
    dy: i64,
    weight: f64,
}

const FLOYD_STEINBERG: &[KernelEntry] = &[
    KernelEntry { dx: 1, dy: 0, weight: 7.0 / 16.0 },
    KernelEntry { dx: -1, dy: 1, weight: 3.0 / 16.0 },
    KernelEntry { dx: 0, dy: 1, weight: 5.0 / 16.0 },
    KernelEntry { dx: 1, dy: 1, weight: 1.0 / 16.0 },
];

const ATKINSON: &[KernelEntry] = &[
    KernelEntry { dx: 1, dy: 0, weight: 1.0 / 8.0 },
    KernelEntry { dx: 2, dy: 0, weight: 1.0 / 8.0 },
    KernelEntry { dx: -1, dy: 1, weight: 1.0 / 8.0 },
    KernelEntry { dx: 0, dy: 1, weight: 1.0 / 8.0 },
    KernelEntry { dx: 1, dy: 1, weight: 1.0 / 8.0 },
    KernelEntry { dx: 0, dy: 2, weight: 1.0 / 8.0 },
];

const JARVIS_JUDICE_NINKE: &[KernelEntry] = &[
    KernelEntry { dx: 1, dy: 0, weight: 7.0 / 48.0 },
    KernelEntry { dx: 2, dy: 0, weight: 5.0 / 48.0 },
    KernelEntry { dx: -2, dy: 1, weight: 3.0 / 48.0 },
    KernelEntry { dx: -1, dy: 1, weight: 5.0 / 48.0 },
    KernelEntry { dx: 0, dy: 1, weight: 7.0 / 48.0 },
    KernelEntry { dx: 1, dy: 1, weight: 5.0 / 48.0 },
    KernelEntry { dx: 2, dy: 1, weight: 3.0 / 48.0 },
    KernelEntry { dx: -2, dy: 2, weight: 1.0 / 48.0 },
    KernelEntry { dx: -1, dy: 2, weight: 3.0 / 48.0 },
    KernelEntry { dx: 0, dy: 2, weight: 5.0 / 48.0 },
    KernelEntry { dx: 1, dy: 2, weight: 3.0 / 48.0 },
    KernelEntry { dx: 2, dy: 2, weight: 1.0 / 48.0 },
];

const STUCKI: &[KernelEntry] = &[
    KernelEntry { dx: 1, dy: 0, weight: 8.0 / 42.0 },
    KernelEntry { dx: 2, dy: 0, weight: 4.0 / 42.0 },
    KernelEntry { dx: -2, dy: 1, weight: 2.0 / 42.0 },
    KernelEntry { dx: -1, dy: 1, weight: 4.0 / 42.0 },
    KernelEntry { dx: 0, dy: 1, weight: 8.0 / 42.0 },
    KernelEntry { dx: 1, dy: 1, weight: 4.0 / 42.0 },
    KernelEntry { dx: 2, dy: 1, weight: 2.0 / 42.0 },
    KernelEntry { dx: -2, dy: 2, weight: 1.0 / 42.0 },
    KernelEntry { dx: -1, dy: 2, weight: 2.0 / 42.0 },
    KernelEntry { dx: 0, dy: 2, weight: 4.0 / 42.0 },
    KernelEntry { dx: 1, dy: 2, weight: 2.0 / 42.0 },
    KernelEntry { dx: 2, dy: 2, weight: 1.0 / 42.0 },
];

struct OrderedState {
    strength: f64,
    alpha_cutoff: f64,
    cell: f64,
    scale: f64,
    seed: u32,
}

/// Apply the dither effect in place — port of `applyDither`.
pub fn apply(out: &mut [u8], w: u32, h: u32, p: &Params, coord: CoordSpace) {
    if w == 0 || h == 0 {
        return;
    }

    let algorithm = p.s("algorithm", "floyd-steinberg");
    let strength = clamp01(p.f("strength", 1.0));
    let alpha_cutoff = clamp01(p.f("alphaCutoff", 0.0));
    let serpentine = p.b("serpentine", true);
    let seed = (js_round(p.f("seed", 0.0)) as i64) as u32;

    let coord_valid = coord.scale > 0.0;
    let scale = if coord_valid { coord.scale } else { 1.0 };
    let phase_x = if coord_valid { coord.region_x } else { 0.0 };
    let phase_y = if coord_valid { coord.region_y } else { 0.0 };
    let origin_x = if coord_valid { coord.origin_x } else { 0.0 };
    let origin_y = if coord_valid { coord.origin_y } else { 0.0 };
    let resolved = CoordSpace { scale, origin_x, origin_y, region_x: phase_x, region_y: phase_y };

    if algorithm == "bayer" || algorithm == "blue-noise" {
        let st = OrderedState {
            strength,
            alpha_cutoff,
            cell: f64::max(1.0, p.f("cellSize", 1.0)) * scale,
            scale,
            seed,
        };
        apply_ordered(out, w, h, p, &st, resolved);
        return;
    }

    if strength <= 0.0 {
        return;
    }

    // 'none' palette mode = no quantization = no dithering at all. Error
    // diffusion requires a quantization target to diffuse error towards.
    if p.s("paletteMode", "levels") == "none" {
        return;
    }

    let metric = ColorMetric::parse(p.s("metric", "rgb"));
    let mut lookup: Option<PaletteLookup> = None;
    if p.s("paletteMode", "levels") == "custom" {
        let raw = p.rgb_list("colors");
        let colors_f64: Vec<[f64; 3]> =
            raw.iter().map(|c| [c[0] as f64, c[1] as f64, c[2] as f64]).collect();
        let colors = palette_core::sanitize_palette(&colors_f64);
        if !colors.is_empty() {
            lookup = Some(PaletteLookup::build(&colors, metric));
        }
    }
    let levels = js_round(p.f("levels", 4.0)).clamp(1.0, 8.0) as i64;
    let step = 1.0 / ((1i64 << levels) - 1) as f64;

    let kernel: &[KernelEntry] = match algorithm {
        "atkinson" => ATKINSON,
        "jarvis-judice-ninke" | "sierra" => JARVIS_JUDICE_NINKE,
        "stucki" => STUCKI,
        _ => FLOYD_STEINBERG,
    };

    let len = (w * h) as usize;
    let mut err_r = vec![0.0f32; len];
    let mut err_g = vec![0.0f32; len];
    let mut err_b = vec![0.0f32; len];

    for y in 0..h {
        let ltr = !serpentine || y % 2 == 0;
        for sx in 0..w {
            let x = if ltr { sx } else { w - 1 - sx };
            let o = ((y * w + x) * 4) as usize;
            let a = out[o + 3];
            if (a as f64) / 255.0 < alpha_cutoff || a == 0 {
                continue;
            }
            let eo = (y * w + x) as usize;
            let r = clamp01((out[o] as f64) / 255.0 + (err_r[eo] as f64) * strength);
            let g = clamp01((out[o + 1] as f64) / 255.0 + (err_g[eo] as f64) * strength);
            let b = clamp01((out[o + 2] as f64) / 255.0 + (err_b[eo] as f64) * strength);

            let (qr, qg, qb): (f64, f64, f64) = if let Some(lookup) = &lookup {
                let c = lookup.find(
                    js_round(r * 255.0) as u8,
                    js_round(g * 255.0) as u8,
                    js_round(b * 255.0) as u8,
                );
                (c[0] as f64 / 255.0, c[1] as f64 / 255.0, c[2] as f64 / 255.0)
            } else {
                (js_round(r / step) * step, js_round(g / step) * step, js_round(b / step) * step)
            };
            out[o] = clamp_byte(clamp01(qr) * 255.0);
            out[o + 1] = clamp_byte(clamp01(qg) * 255.0);
            out[o + 2] = clamp_byte(clamp01(qb) * 255.0);

            let er = (r - qr) * strength;
            let eg = (g - qg) * strength;
            let eb = (b - qb) * strength;
            for entry in kernel {
                let nx = if ltr { x as i64 + entry.dx } else { x as i64 - entry.dx };
                let ny = y as i64 + entry.dy;
                if nx < 0 || nx >= w as i64 || ny >= h as i64 {
                    continue;
                }
                let ne = (ny * w as i64 + nx) as usize;
                err_r[ne] = (err_r[ne] as f64 + er * entry.weight) as f32;
                err_g[ne] = (err_g[ne] as f64 + eg * entry.weight) as f32;
                err_b[ne] = (err_b[ne] as f64 + eb * entry.weight) as f32;
            }
        }
    }
}

/// Ordered (Bayer / blue-noise) pattern dither — port of `applyOrdered`.
fn apply_ordered(out: &mut [u8], w: u32, h: u32, p: &Params, st: &OrderedState, coord: CoordSpace) {
    let levels = js_round(p.f("levels", 4.0)).clamp(1.0, 8.0) as i64;
    let step = 1.0 / ((1i64 << levels) - 1) as f64;
    let metric = ColorMetric::parse(p.s("metric", "rgb"));
    let mut lookup: Option<PaletteLookup> = None;
    if p.s("paletteMode", "levels") == "custom" {
        let raw = p.rgb_list("colors");
        let colors_f64: Vec<[f64; 3]> =
            raw.iter().map(|c| [c[0] as f64, c[1] as f64, c[2] as f64]).collect();
        let colors = palette_core::sanitize_palette(&colors_f64);
        if !colors.is_empty() {
            lookup = Some(PaletteLookup::build(&colors, metric));
        }
    }
    let bayer = if p.s("algorithm", "floyd-steinberg") == "bayer" {
        Some(bayer_threshold_matrix(p.f("bayerSize", 4.0).clamp(2.0, 8.0)))
    } else {
        None
    };

    let inv_scale = 1.0 / st.scale;
    let cell_size_doc = st.cell * inv_scale;

    for y in 0..h {
        for x in 0..w {
            let o = ((y * w + x) * 4) as usize;
            let a = out[o + 3];
            if (a as f64) / 255.0 < st.alpha_cutoff || a == 0 {
                continue;
            }
            // Document-anchored phase: the cell index derives from the
            // document coordinate of the pixel.
            let (doc_x, doc_y) = doc_coord_of(x as f64, y as f64, coord);
            let cell_idx_x = (doc_x / cell_size_doc).floor() as i64;
            let cell_idx_y = (doc_y / cell_size_doc).floor() as i64;
            let threshold: f64;
            if let Some(bayer) = &bayer {
                let size = (bayer.len() as f64).sqrt();
                let cix = cell_idx_x as f64 % size;
                let ciy = cell_idx_y as f64 % size;
                let idx = ciy * size + cix;
                // Out-of-range (negative cell index) yields `undefined` in JS,
                // which propagates NaN through the threshold math and zeroes
                // the pixel on Uint8ClampedArray assignment.
                let t = if idx < 0.0 || idx >= bayer.len() as f64 {
                    f64::NAN
                } else {
                    (bayer[idx as usize] as f64 + 0.5) / (size * size)
                };
                threshold = (t - 0.5) * st.strength * step * 1.5;
            } else {
                threshold =
                    (hash2(cell_idx_x as i32, cell_idx_y as i32, st.seed) - 0.5) * st.strength * step;
            }
            let r = clamp01((out[o] as f64) / 255.0 + threshold);
            let g = clamp01((out[o + 1] as f64) / 255.0 + threshold);
            let b = clamp01((out[o + 2] as f64) / 255.0 + threshold);
            if let Some(lookup) = &lookup {
                let c = lookup.find(
                    js_round(r * 255.0) as u8,
                    js_round(g * 255.0) as u8,
                    js_round(b * 255.0) as u8,
                );
                out[o] = c[0];
                out[o + 1] = c[1];
                out[o + 2] = c[2];
            } else {
                out[o] = quantize_byte(r, step);
                out[o + 1] = quantize_byte(g, step);
                out[o + 2] = quantize_byte(b, step);
            }
        }
    }
}

/// `Math.round(clamp01(Math.round(v / step) * step) * 255)` — port of
/// `quantizeByte`.
fn quantize_byte(v: f64, step: f64) -> u8 {
    clamp_byte(clamp01(js_round(v / step) * step) * 255.0)
}
